### ISSUE TITLE
[Failing Test] node-kubelet-master (ci-kubernetes-node-kubelet)

### ISSUE THREAD

**Issue by hasheddan** [kubernetes/kubernetes#89892]
**Which jobs are failing**:

`node-kubelet-master (ci-kubernetes-node-kubelet)`

**Which test(s) are failing**:

`Node Tests`

**Since when has it been failing**:

`04-06-20 3:44 PDT`

**Testgrid link**:

https://k8s-testgrid.appspot.com/sig-release-master-blocking#node-kubelet-master

**Reason for failure**:

```
W0406 13:57:01.106] 2020/04/06 13:57:01 main.go:314: Something went wrong: encountered 1 errors: [error during go run /go/src/k8s.io/kubernetes/test/e2e_node/runner/remote/run_remote.go --cleanup --logtostderr --vmodule=*=4 --ssh-env=gce --results-dir=/workspace/_artifacts --project=k8s-jkns-ci-node-e2e --zone=us-west1-b --ssh-user=prow --ssh-key=/workspace/.ssh/google_compute_engine --ginkgo-flags=--nodes=8 --focus="\[NodeConformance\]" --skip="\[Flaky\]|\[Serial\]" --test_args=--kubelet-flags="--cgroups-per-qos=true --cgroup-root=/" --test-timeout=1h5m0s --image-config-file=/workspace/test-infra/jobs/e2e_node/image-config.yaml (interrupted): exit status 1]
W0406 13:57:01.110] Traceback (most recent call last):
W0406 13:57:01.110]   File "/workspace/./test-infra/jenkins/../scenarios/kubernetes_e2e.py", line 779, in <module>
W0406 13:57:01.110]     main(parse_args())
W0406 13:57:01.110]   File "/workspace/./test-infra/jenkins/../scenarios/kubernetes_e2e.py", line 627, in main
W0406 13:57:01.111]     mode.start(runner_args)
W0406 13:57:01.111]   File "/workspace/./test-infra/jenkins/../scenarios/kubernetes_e2e.py", line 262, in start
W0406 13:57:01.111]     check_env(env, self.command, *args)
W0406 13:57:01.111]   File "/workspace/./test-infra/jenkins/../scenarios/kubernetes_e2e.py", line 111, in check_env
W0406 13:57:01.111]     subprocess.check_call(cmd, env=env)
W0406 13:57:01.111]   File "/usr/lib/python2.7/subprocess.py", line 190, in check_call
W0406 13:57:01.112]     raise CalledProcessError(retcode, cmd)
W0406 13:57:01.112] subprocess.CalledProcessError: Command '('kubetest', '--dump=/workspace/_artifacts', '--gcp-service-account=/etc/service-account/service-account.json', '--up', '--down', '--test', '--deployment=node', '--provider=gce', '--cluster=bootstrap-e2e', '--gcp-network=bootstrap-e2e', '--gcp-project=k8s-jkns-ci-node-e2e', '--gcp-zone=us-west1-b', '--node-args=--image-config-file=/workspace/test-infra/jobs/e2e_node/image-config.yaml', '--node-test-args=--kubelet-flags="--cgroups-per-qos=true --cgroup-root=/"', '--node-tests=true', '--test_args=--nodes=8 --focus="\\[NodeConformance\\]" --skip="\\[Flaky\\]|\\[Serial\\]"', '--timeout=65m')' returned non-zero exit status 1
```

**Anything else we need to know**:

/cc @kubernetes/ci-signal
/priority critical-urgent
/milestone v1.19
/sig node

**Comment by tedyu** [comment]
From https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-node-kubelet/1247299900119453700/build-log.txt:
```
I0406 23:10:31.875] unable to create gce instance with running docker daemon for image ubuntu-gke-1804-d1703-0-v20181113.  could not create instance tmp-node-e2e-3f80b00f-ubuntu-gke-1804-d1703-0-v20181113: [&{Code:QUOTA_EXCEEDED Location: Message:Quota 'IN_USE_ADDRESSES' exceeded.  Limit: 100.0 in region us-west1. ForceSendFields:[] NullFields:[]}]
```
The cause was due to quota exceeded.

**Comment by yujuhong** [comment]
I checked the project and almost all of the VM instances were created around April 6. Most likely some tests hit timeout and the VMs weren't cleaned up.

I deleted all but one VM in case anyone wanted to check the timeout issue.

**Comment by hasheddan** [comment]
Thanks @yujuhong! I will monitor if anything changes :)

**Comment by yujuhong** [comment]
Failing but for a different reason. Probably related to https://github.com/kubernetes/kubernetes/issues/89847#issuecomment-610109930

**Comment by hasheddan** [comment]
@dchen1107 do you have any insight on this test?

**Comment by spiffxp** [comment]
I just did the same thing @yujuhong did, but I suspect it won't fix anything.  I cleared out VM's so the error you'll see for the next few runs probably won't be quota-related, maybe you can use that to determine if it's the same error as the linked issue.

**Comment by hasheddan** [comment]
Thanks @spiffxp! I'll monitor and follow up :+1:

**Comment by spiffxp** [comment]
Quota exhausted within an hour. It's unclear whether this is a thundering herd problem, or if there's some underlying problem that's causing these tests to either take longer than usual, and/or fail to clean up.

**Comment by hasheddan** [comment]
It appears all tests using this project are experiencing the quota issue (https://testgrid.k8s.io/sig-node-kubelet). However, it appears that the issue may be stemming from the fact that on timeout the deferred cleanup is not being called [here](https://github.com/kubernetes/kubernetes/blob/864b4bcc1e595917f483b5f23451072a679214db/test/e2e_node/runner/remote/run_remote.go#L536). Since there has not been recent changes to this test, I am guessing that something caused the tests to take longer, triggering the timeouts, and then this faulty cleanup on timeout behavior was observed for the first time. It seems like the solution here would be to fix the cleanup behavior, then potentially investigate if the timeout needs to be increased on these tests.

For context on how this gets run (for other folks investigating), here are results of my spelunking:
- Identified [failure](https://testgrid.k8s.io/sig-release-master-blocking#node-kubelet-master) on `sig-release-master-blocking` testgrid dashboard
- Found [job run that had a timeout](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-node-kubelet/1249877524796674048) rather than the quota exception
- Took a look at the [prow job yaml](https://prow.k8s.io/prowjob?prowjob=33c68c1c-7df2-11ea-9571-a25010ecc175) and identified [where it resides in test-infra](https://github.com/kubernetes/test-infra/blob/f468ec6bd93cb672483a96a4019df91ff95e3734/config/jobs/kubernetes/sig-node/node-kubelet.yaml#L1)
- Looked at job logs to identify [where the error was happening](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-node-kubelet/1249877524796674048#1:build-log.txt%3A373)
- Followed the execution flow:
  - This job uses [kubekins-e2e image](https://github.com/kubernetes/test-infra/blob/f468ec6bd93cb672483a96a4019df91ff95e3734/config/jobs/kubernetes/sig-node/node-kubelet.yaml#L16)
  - The [kubekins-e2e image](https://github.com/kubernetes/test-infra/blob/master/images/kubekins-e2e/Dockerfile) is build from the [bootstrap image](https://github.com/kubernetes/test-infra/blob/f468ec6bd93cb672483a96a4019df91ff95e3734/images/kubekins-e2e/Dockerfile#L20)
  - The [bootstrap image](https://github.com/kubernetes/test-infra/blob/master/images/bootstrap/Dockerfile) uses [bootstrap.py as its entrypoint](https://github.com/kubernetes/test-infra/blob/f468ec6bd93cb672483a96a4019df91ff95e3734/images/bootstrap/entrypoint.sh#L25)
  - [bootstrap.py]() checks [scenario](https://github.com/kubernetes/test-infra/blob/f468ec6bd93cb672483a96a4019df91ff95e3734/jenkins/bootstrap.py#L1106) and [invokes scenarios/kubernetes_e2e.py](https://github.com/kubernetes/test-infra/blob/f468ec6bd93cb672483a96a4019df91ff95e3734/jenkins/bootstrap.py#L908) per [job config](https://github.com/kubernetes/test-infra/blob/f468ec6bd93cb672483a96a4019df91ff95e3734/config/jobs/kubernetes/sig-node/node-kubelet.yaml#L21) (ref in [build log](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-node-kubelet/1249877524796674048#1:build-log.txt%3A56))
  - [kubernetes_e2e.py](https://github.com/kubernetes/test-infra/blob/master/scenarios/kubernetes_e2e.py) then [starts kubetest with appropriate args](https://github.com/kubernetes/test-infra/blob/f468ec6bd93cb672483a96a4019df91ff95e3734/scenarios/kubernetes_e2e.py#L627)
  - kubetest [main.go calls e2e.go](https://github.com/kubernetes/test-infra/blob/f468ec6bd93cb672483a96a4019df91ff95e3734/kubetest/main.go#L396), which [runs tests](https://github.com/kubernetes/test-infra/blob/f468ec6bd93cb672483a96a4019df91ff95e3734/kubetest/e2e.go#L50) by running [node tests](https://github.com/kubernetes/test-infra/blob/f468ec6bd93cb672483a96a4019df91ff95e3734/kubetest/e2e.go#L546) with [passed in args](https://github.com/kubernetes/test-infra/blob/f468ec6bd93cb672483a96a4019df91ff95e3734/kubetest/e2e.go#L565)
  - [This](https://github.com/kubernetes/test-infra/blob/f468ec6bd93cb672483a96a4019df91ff95e3734/kubetest/e2e.go#L584) takes us to [process.go](https://github.com/kubernetes/test-infra/blob/f468ec6bd93cb672483a96a4019df91ff95e3734/kubetest/process/process.go#L142), which finally actually runs the [run_remote.go](https://github.com/kubernetes/test-infra/blob/f468ec6bd93cb672483a96a4019df91ff95e3734/kubetest/e2e.go#L567) which is where sig-node has written the tests in [k/k](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/runner/remote/run_remote.go)
  - All of this brings us to where we want to [delete instances we created](https://github.com/kubernetes/kubernetes/blob/864b4bcc1e595917f483b5f23451072a679214db/test/e2e_node/runner/remote/run_remote.go#L536), which I presume we are not doing when these tests timeout

**Comment by spiffxp** [comment]
Some questions:
- why is run_remote timing out?
- is it possible to alter the job so that whatever is hanging in run_remote times out, and it gets to cleanup before the pod is forcefully terminated due to the bootstrap-specified timeout?
- what is it that requires these tests be pinned to this specific gcp project vs. an arbitrary / boskos-managed gcp project?

**Comment by hasheddan** [comment]
@derekwaynecarr do you have any context on this? Specifically on the usage of a single GCP project?

**Comment by spiffxp** [comment]
What we have done:
- Found equivalent coverage
  - Assumed that k8s-jkns-ci-node-e2e has network issues preventing ssh to nodes under test, thus cause all jobs that use it to fail
  - Determined that presubmit jobs running in k8s-jkns-pr-node-e2e run the same tests, and are green
  - Decided presubmit coverage was enough to unblock patch releases

Mitigation:
- Pinned two jobs to a spare GCP project (k8s-jkns-gke-gci-soak)
  - ci-kubernetes-node-kubelet-1-18 (https://github.com/kubernetes/test-infra/pull/17247)
  - ci-kubernetes-node-kubelet-master (https://github.com/kubernetes/test-infra/pull/17251)
- Saw passes for each job
  - https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-node-kubelet-1-18/1250944125981691908
  - https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-node-kubelet/1250930535660785665
- Next steps:
  - pin remaining release-blocking node e2e jobs to unblocked project
  - give node e2e jobs a small pool of boskos projects in case a network issue hits again, so they don't impact other jobs using boskos
  - try making a gcp project under wg-k8s-infra that folks other than google have access to for debugging purposes

**Comment by spiffxp** [comment]
While looking closer at node e2e jobs we noticed they were using years-old COS images

Bumped to latest LTS COS images https://github.com/kubernetes/test-infra/pull/17245

**Comment by hasheddan** [comment]
/close

Green since https://github.com/kubernetes/test-infra/pull/17349
